File: C++/Binary_tree_zigzag_level_order_traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>>res;
        if(!root) return res;
        queue<TreeNode *> q;
        q.push(root);
        
        bool flag = 1;
        while(!q.empty())
        {
            vector<int> row;
            int row_count = q.size();
            flag = flag ^ 1;
            while(row_count)
            {
                TreeNode *t = q.front();
                row.push_back(t->val);
                //if(flag)
                //{
                if(t->left) q.push(t->left);
                if(t->right) q.push(t->right);
                //}
                /*else
                {
                    if(t->right) q.push(t->right);
                    if(t->left) q.push(t->left);
                }*/
                q.pop();
                row_count--;
            }
            if(flag) 
                reverse(row.begin(), row.end());
            res.push_back(row);
        }
        
        return res;
    }
};

File: leetcode/BFS/Binary Tree Zigzag Level Order Traversal.go
package BFS


/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func zigzagLevelOrder(root *TreeNode) [][]int {
	res := [][]int{}
	if root == nil{
		return res
	}
	q := []*TreeNode{}
	q1 := []*TreeNode{}
	q = append(q, root)
	level := 1
	for;len(q) > 0;{
		q1 = []*TreeNode{}
		tmp := []int{}
		for;len(q) > 0;{
			n := q[0]
			q = q[1:]
			tmp = append(tmp, n.Val)
			if n.Left != nil{
				q1  = append(q1, n.Left)
			}
			if n.Right != nil{
				q1 = append(q1, n.Right)
			}
		}
		q = q1
		if level % 2 == 1{
			res = append(res, tmp)
		}else{
			reverse(tmp)
			res = append(res, tmp)
		}
		level++
	}
	return res
}

func reverse(data []int){
	i, j := 0, len(data)-1
	for;i < j;{
		data[i], data[j] = data[j], data[i]
		i++
		j--
	}
}